core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -5854,11 +5860,17 @@ private OutputColumn createOutputColumn(Field field)
         private RedirectionAwareTableHandle getTableHandle(Table table, QualifiedObjectName name, Optional<Scope> scope)
         {
             if (table.getQueryPeriod().isPresent()) {
+                verify(table.getBranch().isEmpty(), "branch must be empty");
                 Optional<TableVersion> startVersion = extractTableVersion(table, table.getQueryPeriod().get().getStart(), scope);
                 Optional<TableVersion> endVersion = extractTableVersion(table, table.getQueryPeriod().get().getEnd(), scope);
                 return metadata.getRedirectionAwareTableHandle(session, name, startVersion, endVersion);
             }
-            return metadata.getRedirectionAwareTableHandle(session, name);
+            if (table.getBranch().isPresent()) {
+                verify(table.getQueryPeriod().isEmpty(), "query period must be empty");
+                Optional<TableVersion> endVersion = table.getBranch().map(branch -> new TableVersion(TARGET_ID, VARCHAR, utf8Slice(branch)));
+                return metadata.getRedirectionAwareTableHandle(session, name, Optional.empty(), endVersion);
```

Comment:
I think we should add another parameter instead of reusing the existing end-version.